app/uk/gov/hmrc/gform/gform/SectionRenderingService.scala

```diff
@@ -476,8 +510,60 @@ class SectionRenderingService(
         htmlForFileUpload(fieldValue, formTemplateId, index, ei, data, userDetails, maybeValidated, lang)
       case InformationMessage(infoType, infoText) =>
         htmlForInformationMessage(fieldValue, infoType, infoText, index, ei)
+      case HmrcTaxPeriod(idType, idNumber, regimeType) =>
+        htmlForHmrcTaxPeriod(
+          fieldValue,
+          index,
+          ei,
+          maybeValidated,
+          data,
+          obligations,
+          HmrcTaxPeriod(idType, idNumber, regimeType))
     }
 
+  private def htmlForHmrcTaxPeriod(
+    fieldValue: FormComponent,
+    index: Int,
+    ei: ExtraInfo,
+    validatedType: ValidatedType,
+    data: FormDataRecalculated,
+    obligations: Map[HmrcTaxPeriod, TaxPeriods],
+    HmrcTP: HmrcTaxPeriod) = {
+
+    val taxPeriodList = obligations
+      .get(HmrcTP) match {
+      case Some(c) => c.taxPeriods
+      case _       => List[TaxPeriod]()
+    }
+
+    val taxPeriodOptions = taxPeriodList
+      .map(
+        i =>
+          new OptionParams(
+            i.periodKey,
+            formatDate(i.inboundCorrespondenceFromDate),
+            formatDate(i.inboundCorrespondenceToDate),
+            false))
+
+    val validatedValue = buildFormFieldValidationResult(fieldValue, ei, validatedType, data)
+    val mapOfResultsOption = validatedValue match {
+      case Some(x) => x
+      case None    => ""
+    }
+    val mapOfResults = mapOfResultsOption match {
+      case ComponentField(a, b) => b
+      case _                    => Map("" -> "")
+    }
+    val setValue = mapOfResults.values.toList match {
+      case a if a.size < 2                    => ""
+      case b: List[FormFieldValidationResult] => TextFormatter.formatText(Some(b(1))).dropRight(1)
```

Comment:
This line produces a warning:

```
[warn]  [E1] app/uk/gov/hmrc/gform/gform/SectionRenderingService.scala:560:15:
[warn]       non-variable type argument uk.gov.hmrc.gform.validation.FormFieldValidationResult in type pattern List[uk.gov.hmrc.gform.validation.FormFieldValidationResult] (the underlying of List[uk.gov.hmrc.gform.validation.FormFieldValidationResult]) is unchecked since it is eliminated by erasure
[warn]             case b: List[FormFieldValidationResult] =>
```
It can be rewritten with a use of constructors based pattern matching to get rid of that warning.